### PR TITLE
UNOMI-486 JSON Schema integration

### PR DESCRIPTION
**Pull request by sergehuber**
This PR is the beginning of the integration of JSON schema to validate event types and other incoming objects.

This PR contains the following:
- A new SchemaRegistry that is responsible for loading the declared schemas in memory (in the future we will probably also want to store them in ES) and making them available.
- Predefined schemas for all the event types and value types known in Unomi.
- A first integration of the schema validation in the EventServiceImpl.isEventValid method, but that isn't blocking right now until the system is properly reviewed.
- Implementation of a PropertyType keyword that makes it possible to extension the JSON schema validation to validate property types defined at runtime. This is a PoC right now but it shows how the validation library can be extended to fit any validation need.

Note that this implementation will make the EventTypeRegistry obsolete once we have iso-functionality for the GraphQL implementation but it is kept for the moment as a transition to make sure the system can still be tested. To get this to work we will need to access the SchemaType structures to generate the GraphQL schema from the JSON schema definitions.

### CODE REVIEW

**Review by jkevan — APPROVED** [review]
Nothing to mention, others than what is already raised in other comments.
Great work.

File: services/src/main/java/org/apache/unomi/services/impl/schemas/SchemaRegistryImpl.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        return schemaTypesById.get(schemaId);
+    }
+
+    private void loadPredefinedSchemas(BundleContext bundleContext) {
+        Enumeration<URL> predefinedSchemas = bundleContext.getBundle().findEntries("META-INF/cxs/schemas", "*.json", true);
+        if (predefinedSchemas == null) {
+            return;
+        }
+
+        ObjectMapper objectMapper = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (sergehuber) [reply services/src/main/java/org/apache/unomi/services/impl/schemas/SchemaRegistryImpl.java]: I've done this change. Thanks

File: services/src/main/java/org/apache/unomi/services/impl/schemas/SchemaRegistryImpl.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+        List<SchemaType> schemaTypes = this.schemaTypesByBundle.get(bundleContext.getBundle().getBundleId());
+
+        while (predefinedSchemas.hasMoreElements()) {
+            URL predefinedSchemaURL = predefinedSchemas.nextElement();
+            logger.debug("Found predefined JSON schema at " + predefinedSchemaURL + ", loading... ");
+
+            try {
```
[the remark this answers is not in the document]
Comment (sergehuber) [reply services/src/main/java/org/apache/unomi/services/impl/schemas/SchemaRegistryImpl.java]: Thanks, I've changed to use try-with-resources.

File: services/src/main/java/org/apache/unomi/services/impl/schemas/SchemaRegistryImpl.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+    public void setProfileService(ProfileService profileService) {
+        this.profileService = profileService;
+    }
+
+    public boolean isValid(Object object, String schemaId) {
+        JsonSchema jsonSchema = jsonSchemasById.get(schemaId);
+        if (jsonSchema != null) {
+            try {
+                // this is a workaround to get the JsonNode from the object, maybe there is a better way (more efficient) to do this ?
```
[the remark this answers is not in the document]
Comment (sergehuber) [reply services/src/main/java/org/apache/unomi/services/impl/schemas/SchemaRegistryImpl.java]: Thanks, I've changed to use this method.

**Pull request merged**